### PR TITLE
ASCAT mask maker input utility program

### PR DESCRIPTION
**Pull request by amfox37**
This program reads in a NetCDF file containing ASCAT soil moisture masks available from:
Lindorfer, R., Wagner, W., Hahn, S., Kim, H., Vreugdenhil, M., Gruber, A., Fischer, M., & Trnka, M. (2023). 
Global Scale Maps of Subsurface Scattering Signals Impacting ASCAT Soil Moisture Retrievals (1.0.0) [Data set]. 
TU Wien. https://doi.org/10.48436/9a2y9-e5z14

It provides the possibility to combine different masks (current default case is combination of subsurface and wetland masks) and interpolates onto a regular grid with a (hardwired) 0.1 degree lat/lon resolution and -90/-180 degree lower left corner used for quick indexing in the ASCAT observation reader QC routine, writing the final output to a NetCDF file.

### CODE REVIEW

File: src/Applications/LDAS_App/CMakeLists.txt [hunk 1]
```diff
@@ -13,6 +13,12 @@ ecbuild_add_executable (
    TARGET tile_bin2nc4.x 
    SOURCES tile_bin2nc4.F90
    LIBS MAPL)
+ 
+ecbuild_add_executable (
+   TARGET ascat_mask_maker.x
+   SOURCES util/inputs/ASCAT_sm_mask/ascat_mask_maker.F90
+   LIBS MAPL
+)
```
Comment (gmao-rreichle) [inline src/Applications/LDAS_App/CMakeLists.txt:21]: @weiyuan-jiang : Should we have a CMakeLists.txt file in the sub-dir?  Or is it ok to have only the CMakeLists.txt in ./LDAS_App ?  Also, can you please make sure that the new executable is built but **not** installed?  Thanks!

File: src/Applications/LDAS_App/util/inputs/ASCAT_sm_mask/ascat_mask_maker.F90 [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+    character(400)                  :: fname_in
+
+    ! --------------------------------------------------------------------------------
+    !
+    ! hardwired variables
+
+    ! ASCAT soil moisture mask file from Lindorfer et al 2023 
+    
+    fname_in = '/discover/nobackup/amfox/subsurface_scattering_ASCAT_ERA5_Land.nc'
```
Comment (gmao-rreichle) [inline src/Applications/LDAS_App/util/inputs/ASCAT_sm_mask/ascat_mask_maker.F90:36]: @amfox37, @biljanaorescanin : We'll need to find a project disk where we store this type of input file.  It should not live in a user's directory.

File: src/Applications/LDAS_App/util/inputs/ASCAT_sm_mask/ascat_mask_maker.F90 [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+       combined_mask = subsurface_mask
+    end where
+
+    ! Re-map "combined_mask" from WARP5 input grid to regular lat/lon output grid (2-dim array)
+
+    allocate(lon(int(360.0  / d_lon)))
+    allocate(lat(int(180.0  / d_lat)))
+
+    lon = [(ll_lon + i * d_lon, i = 0, size(lon) - 1)]
+    lat = [(ll_lat + i * d_lat, i = 0, size(lat) - 1)]
```
Comment (gmao-rreichle) [inline src/Applications/LDAS_App/util/inputs/ASCAT_sm_mask/ascat_mask_maker.F90:106]: @amfox37 :  What is "lon" and "lat" exactly?  A quick glance at the ASCAT_EUMETSAT reader suggests that "lon" and "lat" are the western/southern boundaries of the grid cell for which the corresponding i,j mask value applies.  Without this cross-referencing, I would have thought that the "lon" and "lat" values here are for a dateline-on-center and pole-on-center grid.  Please double-check to make sure everything is consistent and also documented in the code.

File: src/Applications/LDAS_App/util/inputs/ASCAT_sm_mask/ascat_mask_maker.F90 [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+    implicit none
+
+    integer :: ncid, varid, dimid, ierr, len, N_gpi, dimids(2)
+    integer :: i, j, closest_index
+
+    integer(kind=1), dimension(:,:), allocatable :: mask_out
+    integer(kind=1)                              :: missing_value
+
+    real, dimension(:), allocatable :: asc_lon, asc_lat, cold_mask, wet_mask, veg_mask, subsurface_mask, combined_mask
```
Comment (gmao-rreichle) [inline src/Applications/LDAS_App/util/inputs/ASCAT_sm_mask/ascat_mask_maker.F90:24]: A brief look at the nc file suggests that the *_mask variables are stored as int8 in the original nc file.  Do we have to read them as real here and put them into the real array `combined_mask`?  The `where (wet_mask == 1)` condition below is risky for real numbers.  My sense is that everything should be in integer space if possible.

Comment (amfox37) [reply src/Applications/LDAS_App/util/inputs/ASCAT_sm_mask/ascat_mask_maker.F90:24]: Yes, all the mask variables should be integers. I'll change it.

File: src/Applications/LDAS_App/util/inputs/ASCAT_sm_mask/ascat_mask_maker.F90 (hunk 3 above)
Comment (amfox37) [reply src/Applications/LDAS_App/util/inputs/ASCAT_sm_mask/ascat_mask_maker.F90:106]: Hmm... I'd written this so lat/lon were the western/southern boundaries to correspond with the reader - but I realize now that's probably not what we want as these values are actually being used to calculate the nearest neighbor. And in that case we need the lat/lon for the middle of the grid cell I think?

**Comment by gmao-rreichle** [comment]
@amfox37 : Thanks for the latest changes.  I added [one more commit](https://github.com/GEOS-ESM/GEOSldas/pull/729/commits/5805a55f9b74e299880bd7deff97fddfa958a3c7).  Please take a look and let me know if the changes are ok with you.

File: src/Applications/LDAS_App/util/inputs/ASCAT_sm_mask/ascat_mask_maker.F90 (hunk 2 above)
Comment (gmao-rreichle) [reply src/Applications/LDAS_App/util/inputs/ASCAT_sm_mask/ascat_mask_maker.F90]: @amfox37 :  For lack of a better idea, I suggest we add the Lindorfer et al 2023 mask file into the ASCAT EUMETSAT obs dir.   But I don't know where the files are.  The data dir from the `LDASsa_DEFAULT_inputs_ensupd.nml` file does not seem to exist, which would mean that we also need to fix the nml file:
`/discover/nobackup/projects/gmao/merra/iau/merra_land/DATA/ASCAT_EUMETSAT/`
cc: @gmao-qliu @biljanaorescanin

Comment (amfox37) [reply src/Applications/LDAS_App/util/inputs/ASCAT_sm_mask/ascat_mask_maker.F90]: The observations are here:
/discover/nobackup/projects/gmao/smap/SMAP_Nature/ASCAT_EUMETSAT

In the past we discussed if this was a suitable home - and concluded it was OK for now. I can update the default namelists if that is still the case.

File: src/Applications/LDAS_App/CMakeLists.txt (hunk 1 above)
Comment (amfox37) [reply src/Applications/LDAS_App/CMakeLists.txt:21]: Here I followed the example for mwrtm_bin2nc4.x on the next few lines which is:
`ecbuild_add_executable (
   TARGET mwrtm_bin2nc4.x
   SOURCES util/inputs/mwRTM_params/mwrtm_bin2nc4.F90
   LIBS GEOSlandassim_GridComp)`

Comment (weiyuan-jiang) [reply src/Applications/LDAS_App/CMakeLists.txt:21]: > @weiyuan-jiang : Any thoughts on the above? I'd like to merge this PR soon.
Yes, we can make it build but not install
If there is no other fortran files, we dont need to add all the CMakeLists.txt

**Comment by gmao-rreichle** [comment]
@amfox37 : I implemented the path/filename changes in [this commit](https://github.com/GEOS-ESM/GEOSldas/pull/729/commits/48c5ece1894ca7a7817c6a924ad628ebaf0f5f79).  Please review, test the new functionality from scratch (new build) to make sure that ascat_mask_maker.x still works as intended, and then let me know.  Once I have your confirmation, I'll go ahead and merge.  
Also, I made a copy of the Lindorfer et al file into a new sub-dir "./Mask", so please remove: `/discover/nobackup/projects/gmao/smap/SMAP_Nature/ASCAT_EUMETSAT/subsurface_scattering_ASCAT_ERA5_Land.nc`

File: src/Applications/LDAS_App/CMakeLists.txt
Comment (amfox37) [inline src/Applications/LDAS_App/CMakeLists.txt:1]: @weiyuan-jiang @gmao-rreichle My update to the CMakeLists works OK and builds but doesn't install, but I don't know if adding`esma_add_subdirectories(util/inputs/ASCAT_sm_mask)` is the correct thing to do?

**Comment by amfox37** [comment]
File removed. 
Everything seems to be working as expected with a clean build - although I'm afraid I don't know if adding `esma_add_subdirectories(util/inputs/ASCAT_sm_mask)` is actually the correct thing to do? Or are you supposed to add a file to each subdirectory and only add one level at a time? Otherwise good to go I think

**Comment by gmao-rreichle** [comment]
> File removed. Everything seems to be working as expected with a clean build - although I'm afraid I don't know if adding `esma_add_subdirectories(util/inputs/ASCAT_sm_mask)` is actually the correct thing to do? Or are you supposed to add a file to each subdirectory and only add one level at a time? Otherwise good to go I think

Thanks, @amfox37.  I think @weiyuan-jiang gave his ok on the single CMakeLists.txt file [here](https://github.com/GEOS-ESM/GEOSldas/pull/729#discussion_r1530791156): "_If there is no other fortran files, we dont need to add all the CMakeLists.txt_" (note that the formatting in the original comment was a bit messed up).  
I will go ahead and ask for official CMake approval.

**Review by gmao-rreichle — APPROVED** [review]

**Comment by gmao-rreichle** [comment]
@mathomp4, @weiyuan-jiang: FYI, I was able to merge this GEOSldas PR **without** formal CMake approval.  Because @weiyuan-jiang had already looked at it, I went for it and merged.

**Pull request merged**